### PR TITLE
#203 Using SAT to migrate models(Take into account the DefaultDomain of the included Person objects)

### PR DESCRIPTION
**Pull request by adrianchivu**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-SDK-Community-Edition/pulls) open
- [X] I have verified that I am following the COMET-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-SDK-Community-Edition/master/.github/CONTRIBUTING.md)
- [X] I have provided test coverage for my change (where applicable)

### Description
Make the pruning algorithm (in either JsonFileDalUtils.AddDomainsOfExpertise or JsonFileDalUtils.AddPersons) also take into account the DefaultDomain of the included Person object.

### CODE REVIEW

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 1]
```diff
@@ -210,7 +210,7 @@ public override Task<IEnumerable<Thing>> Write(IEnumerable<OperationContainer> o
                 JsonFileDalUtils.AddDomainsOfExpertise(engineeringModelSetup, ref domainsOfExpertises);
 
                 // add the Persons that are to be included in the File
-                JsonFileDalUtils.AddPersons(engineeringModelSetup, ref persons, ref personRoles, ref participantRoles, ref organizations);
+                JsonFileDalUtils.AddPersons(engineeringModelSetup, ref persons, ref personRoles, ref participantRoles, ref organizations, ref domainsOfExpertises);
```
Comment (alexatrhea) [inline CDP4JsonFileDal/JsonFileDal.cs:213]: `domainsOfExpertise`

File: CDP4JsonFileDal/JsonFileDalUtils.cs [hunk 2]
```diff
@@ -172,12 +172,16 @@ internal static void AddDomainsOfExpertise(EngineeringModelSetup engineeringMode
...
+        /// The target <see cref="HashSet{DomainOfExpertise}"/>
+        /// </param>
         internal static void AddPersons(
             EngineeringModelSetup engineeringModelSetup,
             ref HashSet<Person> persons,
             ref HashSet<PersonRole> personRoles,
             ref HashSet<ParticipantRole> participantRoles,
-            ref HashSet<Organization> organizations)
+            ref HashSet<Organization> organizations,
+            ref HashSet<DomainOfExpertise> domainOfExpertises)
```
Comment (alexatrhea) [inline CDP4JsonFileDal/JsonFileDalUtils.cs:184]: `domainsOfExpertise`

**Review by alexatrhea — CHANGES_REQUESTED** [review]
A more correct conjugation.

**Review by cozminvelciu — APPROVED** [review]

**Comment by cozminvelciu** [comment]
Fixes https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues/203

**Review by alexatrhea — APPROVED** [review]

File: CDP4JsonFileDal/JsonFileDalUtils.cs [hunk 3]
```diff
@@ -140,16 +140,33 @@ internal static void AddIterationSetup(IterationSetup iterationSetup, ref HashSe
...
+            {
+                foreach (var domainOfExpertise in participant.Domain)
                 {
-                    domainOfExpertises.Add(domainOfExpertise);
+                    if (!domainsOfExpertise.Contains(domainOfExpertise))
+                    {
+                        domainsOfExpertise.Add(domainOfExpertise);
+                    }
+
+                    if (participant.Person.DefaultDomain != null &&
```
Comment (cozminvelciu) [inline CDP4JsonFileDal/JsonFileDalUtils.cs:165]: This `if` should be outside the `domainOfExpertise` `foreach`.

**Review by cozminvelciu — CHANGES_REQUESTED** [review]

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 4]
```diff
@@ -176,7 +176,7 @@ public override Task<IEnumerable<Thing>> Write(IEnumerable<OperationContainer> o
             var iterations = new HashSet<CDP4Common.EngineeringModelData.Iteration>();
             var siteReferenceDataLibraries = new HashSet<CDP4Common.SiteDirectoryData.SiteReferenceDataLibrary>();
             var modelReferenceDataLibraries = new HashSet<CDP4Common.SiteDirectoryData.ModelReferenceDataLibrary>();
-            var domainsOfExpertises = new HashSet<CDP4Common.SiteDirectoryData.DomainOfExpertise>();
+            var domainsOfExpertise = new HashSet<CDP4Common.SiteDirectoryData.DomainOfExpertise>();
```
Comment (lxatrhea) [inline CDP4JsonFileDal/JsonFileDal.cs:179]: The plural version of "domain of expertise" is indeed "domain**s** of expertise".
From a readability point of view I find domainOfExpertise**s** a better choice as a variable name. 
I noticed that I had more problems reading the AddDomainsOfExpertise method. I somehow read over the 's' in domains. The difference between singural and plural seems easier when the plural version ends with an 's'.

File: CDP4JsonFileDal/JsonFileDalUtils.cs (hunk 3 above)
Comment (adrianchivu) [reply CDP4JsonFileDal/JsonFileDalUtils.cs:165]: Correct, copy/paste error.

**Review by cozminvelciu — APPROVED** [review]

File: CDP4JsonFileDal/JsonFileDal.cs (hunk 4 above)
Comment (lxatrhea) [reply CDP4JsonFileDal/JsonFileDal.cs]: This is not what I meant. Please keep the **s** at the end of the variable name.

**Review by cozminvelciu — APPROVED** [review]

**Pull request merged**